src/chains/ethereum/ethereum/tests/api/eth/logs.test.ts

```diff
@@ -33,20 +29,37 @@ describe("api", () => {
           "eth_getTransactionReceipt",
           [transactionHash]
         );
-        contractAddress = transactionReceipt.contractAddress;
         await provider.send("eth_unsubscribe", [subscriptionId]);
+        return transactionReceipt.contractAddress;
+      };
+
+      beforeEach(async () => {
+        contract = compile(join(__dirname, "./contracts/Logs.sol"));
+        provider = await getProvider();
+        accounts = await provider.send("eth_accounts");
+
+        contractAddress = await deployContractAndGetAddress();
       });
 
       describe("eth_subscribe", () => {
         describe("logs", () => {
-          const onceMessageFor = subId => {
-            return new Promise<any>(resolve => {
+          const getMessagesForSub = async (
+            subId,
+            expectedMessageCount,
+            messages
+          ) => {
+            await new Promise<any>(resolve => {
               provider.on("message", message => {
                 if (message.data.subscription === subId) {
+                  messages.push(message);
+                  if (expectedMessageCount > messages.length + 1) {
+                    getMessagesForSub(subId, expectedMessageCount, messages);
+                  }
                   resolve(message);
                 }
               });
             });
+            return messages;
```

Comment:
The recursion seems strange here. Something like the following should do the same job and I think it seems a bit simpler. Thoughts?

```typescript
          const getMessagesForSub = async (
            subId,
            expectedMessageCount
          ) => {
            const messages = [];
            return await new Promise<any[]>(resolve => {
              provider.on("message", message => {
                if (message.data.subscription === subId) {
                  messages.push(message);
                  if (expectedMessageCount === messages.length) {
                    resolve(messages);
                  }
                }
              });
            });
```

Reply:
Yours is waaaaaay better, fixing.